File: backend/NEOBTE/src/main/java/com/sesame/neobte/Services/Loan/LoanService.java
package com.sesame.neobte.Services.Loan;

import com.sesame.neobte.DTO.Requests.Loan.LoanProductFormDTO;
import com.sesame.neobte.DTO.Requests.Loan.LoanRequestDTO;
import com.sesame.neobte.DTO.Responses.Loan.LoanProductResponseDTO;
import com.sesame.neobte.DTO.Responses.Loan.LoanResponseDTO;

import java.util.List;

public interface LoanService {
    // Products (admin)
    List<LoanProductResponseDTO> getAllProducts();
    List<LoanProductResponseDTO> getActiveProducts();
    LoanProductResponseDTO createProduct(LoanProductFormDTO dto);
    LoanProductResponseDTO updateProduct(Long id, LoanProductFormDTO dto);
    void deleteProduct(Long id);

    // Loans (client)
    LoanResponseDTO requestLoan(LoanRequestDTO dto, Long userId);
    List<LoanResponseDTO> getMyLoans(Long userId);
    LoanResponseDTO getLoanById(Long loanId, Long userId);

    // Admin
    List<LoanResponseDTO> getAllLoans();
    LoanResponseDTO approveLoan(Long loanId, String adminNote);
    LoanResponseDTO rejectLoan(Long loanId, String motif);

    // Schedulers
    void processMonthlyRepayments();
    void applyPenalties();
}

File: frontend/NEOBTE/src/app/Services/loan.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Loan, LoanProduct, LoanProductFormDTO, LoanRequestDTO, LoanSimulation } from '../Entities/Interfaces/loan';

@Injectable({ providedIn: 'root' })
export class LoanService {
    private clientApi = 'http://localhost:8080/api/v1/client/loans';
    private adminApi = 'http://localhost:8080/api/v1/admin/loans';
    constructor(private http: HttpClient) { }

    // Client
    getProducts(): Observable<LoanProduct[]> {
        return this.http.get<LoanProduct[]>(`${this.clientApi}/products`);
    }
    simulate(montant: number, tauxAnnuel: number, dureeEnMois: number): Observable<LoanSimulation> {
        return this.http.get<LoanSimulation>(`${this.clientApi}/simulate`, {
            params: { montant, tauxAnnuel, dureeEnMois }
        });
    }
    requestLoan(dto: LoanRequestDTO): Observable<Loan> {
        return this.http.post<Loan>(`${this.clientApi}/request`, dto);
    }
    getMyLoans(): Observable<Loan[]> {
        return this.http.get<Loan[]>(`${this.clientApi}/my`);
    }

    // Admin
    getAllProducts(): Observable<LoanProduct[]> {
        return this.http.get<LoanProduct[]>(`${this.adminApi}/products`);
    }
    createProduct(dto: LoanProductFormDTO): Observable<LoanProduct> {
        return this.http.post<LoanProduct>(`${this.adminApi}/products`, dto);
    }
    updateProduct(id: number, dto: LoanProductFormDTO): Observable<LoanProduct> {
        return this.http.put<LoanProduct>(`${this.adminApi}/products/${id}`, dto);
    }
    deleteProduct(id: number): Observable<any> {
        return this.http.delete(`${this.adminApi}/products/${id}`);
    }
    getAllLoans(): Observable<Loan[]> {
        return this.http.get<Loan[]>(`${this.adminApi}/all`);
    }
    approveLoan(id: number, adminNote: string): Observable<Loan> {
        return this.http.put<Loan>(`${this.adminApi}/${id}/approve`, { adminNote });
    }
    rejectLoan(id: number, motif: string): Observable<Loan> {
        return this.http.put<Loan>(`${this.adminApi}/${id}/reject`, { motif });
    }
}